daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -375,188 +374,218 @@ object ScenarioLedger {
 
   case class UniqueKeyViolation(gk: GlobalKey)
 
-  private def processTransaction(
+  /** Functions for updating the ledger with new transactional information.
+    *
+    * @param trId transaction identity
+    * @param richTr (enriched) transaction
+    * @param locationInfo location map
+    */
+  class TransactionProcessor(
       trId: TransactionId,
       richTr: RichTransaction,
       locationInfo: Map[NodeId, Location],
-      ledgerData: LedgerData,
-  ): Either[UniqueKeyViolation, LedgerData] = {
+  ) {
 
-    final case class RollbackBeginState(
-        activeContracts: Set[ContractId],
-        activeKeys: Map[GlobalKey, ContractId],
-    )
+    def duplicateKeyCheck(ledgerData: LedgerData): Either[UniqueKeyViolation, Unit] = {
+      val inactiveKeys = richTr.transaction.contractKeyInputs
+        .fold(error => crash(s"$error: inconsistent transaction"), identity)
+        .collect { case (key, _: Tx.KeyInactive) =>
+          key
+        }
 
-    final case class ProcessingNode(
-        // The id of the direct parent or None.
-        mbParentId: Option[NodeId],
-        // The id of the nearest rollback ancestor. If this is
-        // a rollback node itself, it points to itself.
-        mbRollbackAncestorId: Option[NodeId],
-        children: List[NodeId],
-        // For rollback nodes, we store the previous state here and restore it.
-        // For exercise nodes, we don’t need to restore anything.
-        prevState: Option[RollbackBeginState],
-    )
+      inactiveKeys.find(ledgerData.activeKeys.contains(_)) match {
+        case Some(duplicateKey) =>
+          Left(UniqueKeyViolation(duplicateKey))
 
-    @tailrec
-    def processNodes(
-        cache0: LedgerData,
-        enps: List[ProcessingNode],
-    ): LedgerData = enps match {
-      case Nil => cache0
-      case ProcessingNode(_, _, Nil, optPrevState) :: restENPs => {
-        val cache1 = optPrevState.fold(cache0) { case prevState =>
-          cache0.copy(
-            activeContracts = prevState.activeContracts,
-            activeKeys = prevState.activeKeys,
-          )
-        }
-        processNodes(cache1, restENPs)
+        case None =>
+          Right(())
+      }
+    }
+
+    def addNewLedgerNodes(historicalLedgerData: LedgerData): LedgerData =
+      richTr.transaction.transaction.reachableNodeIds.foldLeft(historicalLedgerData) {
```

Comment:
Here we can assume the transaction is well formed, meaning all the node Ids are reachable from the root nodes. 

You could use directly `richTr.transaction.transaction.nodes.keys`, but I will favor direct use of `richTr.transaction.transaction.fold`.